tests/test_project.py

```diff
@@ -2517,6 +2531,224 @@ def test_raise_exception_in_hook_cmd(self):
+            with open(project.fn("dirty.txt"), "w"):
+                pass
+            repo.index.add(["dirty.txt"])
+        return repo
+
+    def split_log(self, job, fn):
+        with open(job.fn(fn)) as f:
+            values = f.read().split("\n")
+        if values[-1] == "":
+            values = values[:-1]
+        return values
+
+    def get_log(self, job, fn=None):
+        if fn is None:
+            execution_history = job.doc["execution_history"]()
+            # Add back in redundent information to make check_metadata universal
+            for key, item in execution_history.items():
+                for entry in item:
+                    entry["job-operation"] = {"name": key, "job_id": job.id}
+            return execution_history
+        log_entries = [json.loads(entry) for entry in self.split_log(job, fn)]
+        execution_history = {}
+        for entry in log_entries:
+            operation_name = entry["job-operation"]["name"]
+            execution_history.setdefault(operation_name, []).append(entry)
+        return execution_history
+
+    def check_metadata(
+        self,
+        metadata,
+        job,
+        operation_name,
+        error_message,
+        before_execution,
+        expected_stage,
+        repo,
+    ):
+        assert metadata["stage"] == expected_stage
+        # When running on MacOS CI a private/ is prepended to job.project.path
+        # seemingly indeterminetly. If metadata["project"]["path"] is checked
+        # then job.project.path will have "private" in it. If I check
+        # job.project.path then metadata["project"]["path"] will have "private"
+        # in it. If I check neither, one of them will have "private" in it.
+        test_path = os.path.join(
+            *filter(lambda x: x != "private", Path(metadata["project"]["path"]).parts)
+        )
+        current_path = os.path.join(
+            *filter(lambda x: x != "private", Path(job.project.path).parts)
+        )
+        assert current_path == test_path
+        assert metadata["project"]["schema_version"] == job.project.config.get(
+            "schema_version"
+        )
+        start_time = datetime.datetime.fromisoformat(metadata["time"])
+        assert before_execution < start_time
+        job_op_metadata = metadata["job-operation"]
+        assert job_op_metadata["job_id"] == job.id
+        assert job_op_metadata["name"] == operation_name
+
+        if expected_stage != "prior" and job.sp.raise_exception:
+            assert error_message in metadata["error"]
+
+        else:
+            assert metadata["error"] is None
+
+        if repo is not None:
+            assert metadata["project"]["git"]["commit_id"] == str(repo.commit())
+            assert metadata["project"]["git"]["dirty"] == repo.is_dirty()
+        else:
+            if self.error_on_no_git:
+                # Should not happen in actual testing, here for developers.
+                assert False, "Expected gitpython for test."
```

Comment:
This doesn't seem right. What is "actual testing"? Do you mean CI is "actual testing" but local runs of the test suite are not? It is okay to require that developers have gitpython installed to run the test suite, or you can edit tests requiring gitpython to call `pytest.importorskip`.

Reply:
It isn't testing in any case. It is there to ensure we never reach this point. It should be impossible to get here. If that isn't the case there is an error in the test.